### ISSUE TITLE
MachinePool workers support in ClusterClass

### ISSUE THREAD

**Issue by dlipovetsky** [kubernetes-sigs/cluster-api#5991]
**User Story**

As a user, I would like to create a ClusterClass-based Cluster where workers are managed using MachinePools instead of MachineDeployments. This allows me, for example, to use ClusterClass with an infrastructure provider "flavor" like Cluster API Azure Provider Managed AKS, which supports only MachinePools.

**Detailed Description**

I want to be able to define a ClusterClass where workers are managed using MachinePools instead of MachineDeployments.

```yaml
---
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterClass
metadata:
  name: example
spec:
  controlPlane:
...
  workers:
    machinePools:
      - class: example
        template:
          spec:
            clusterName: example
            replicas: 10
            template:
              spec:
                bootstrap:
                  ref:
                    apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
                    kind: KubeadmConfigTemplate
                    name: example-pool-0
                clusterName: example
                infrastructureRef:
                  apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
                  kind: AzureManagedMachinePool
                  name: example-pool-0
                version: 1.22.4
  infrastructure:
...
```

**Anything else you would like to add:**

This issue follows up on a [slack conversation](https://kubernetes.slack.com/archives/C8TSNPY4T/p1638289221256300?thread_ts=1638288074.252700&cid=C8TSNPY4T).

## Tasks

[EDIT: by @sbueringer] After further discussion we would implement this with the following sub-tasks:

### APIs #8820 

* [x] Extend APIs (api/v1beta1)
  * Cluster
    * WorkersTopology: Add MachinePoolTopology slice
  * ClusterClass
    * WorkersClass: Add MachinePoolClass slice
    * PatchSelectorMatch: Add MachinePool
  * Note: excluding MachineHealthCheck fields

### Controllers, webhooks, ... #9016

* [x] Reconcile MachinePools in Cluster topology (internal/controllers/topology/cluster)
  * Basic reconcile
    * blueprint.go: also get MachinePool blueprint
    * current_state.go: also get MachinePool current state
    * desired_state.go: calculate desired MachinePool like MachineDeployment
    * reconcile_state.go: reconcile MachinePool like MachineDeployment
    * cluster_controller.go: also watch MachinePools
    * Surface MachinePool state like MachineDeployment state in conditions (conditions.go)
  * Support patching MachinePools:
    * Extend patch engine to also handle MachinePools
    * Extend json patch generator to handle MachinePools
    * Add MachinePool builtin variables
* [x] Add corresponding validation in internal/webhooks/
  * Includes additions in internal/topology/check/ & internal/topology/variables/{cluster_variable_defaulting.go, cluster_variable_validation.go}
* [x] Extend internal/controllers/ClusterClass
  * Add the same that we already have for MachineDeployments
* [x] Extend CAPD
  * Add DockerMachinePoolTemplate
* [x] Extend test/extension
  * Also patch DockerMachinePoolTemplate
* [x] Extend quickstart test to also cover MachinePools

### unit & integration tests

Take a look at the existing tests and the code added for MachinePools and add test coverage accordingly for MachinePools.

Cluster topology controller:
  * [x] internal/controllers/topology/cluster/blueprint_test.go #9348 @willie-yao 
  * [x] internal/controllers/topology/cluster/current_state_test.go #9348 @willie-yao 
  * [x] internal/controllers/topology/cluster/desired_state_test.go #9348 @willie-yao 
  * [x] internal/controllers/topology/cluster/reconcile_state_test.go #9380 @willie-yao 
  * [x] internal/controllers/topology/cluster/cluster_controller_test.go #9380 @willie-yao 
  * [x] internal/controllers/topology/cluster/conditions_test.go #9380 @willie-yao 
  * [x] internal/controllers/topology/cluster/scope/scope_test.go @razashahid107
  * [ ] internal/controllers/topology/cluster/scope/state_test.go
  * [ ] internal/controllers/topology/cluster/scope/upgradetracker_test.go

Inline patching:
  * [x] internal/controllers/topology/cluster/patches/inline/json_patch_generator_test.go #9474 @johannesfrey 
  * [x] internal/controllers/topology/cluster/patches/variables/variables_test.go #9372 @johannesfrey 
  * [x] internal/controllers/topology/cluster/patches/engine_test.go #9373 @johannesfrey 

Webhooks:
  * [ ] internal/topology/check/compatibility_test.go
  * [ ] internal/webhooks/cluster_test.go
  * [ ] internal/webhooks/clusterclass_test.go
  * [ ] internal/webhooks/patch_validation_test.go

ClusterClass controller:
  * [ ] internal/controllers/clusterclass/clusterclass_controller_test.go

External patching: test extension:
  * [ ] test/extension/handlers/topologymutation/handler_test.go

clusterctl alpha topology dry-run:
  * [ ] cmd/clusterctl/client/cluster/topology_test.go

Verification:
* [ ] Run unit & integration tests with coverage and check if we missed something

### e2e tests

* First wave:
  * [x] Quickstart (quick_start.go, flavor: "topology,topology-dualstack-.*") #9393 @willie-yao 
  * [x] Kubernetes upgrade (cluster_upgrade.go, flavors: "upgrades,topology,kcp-scale-in") https://github.com/kubernetes-sigs/cluster-api/pull/9502 @willie-yao 
    * The ClusterClass code path should use MPs via CC
* Afterwards:
  * [ ] ClusterClass changes & rollout (clusterclass_changes.go & clusterclass_rollout, flavor: "topology")
    * Ensure assert also verifies the labels/annotations (also in-place rollout of them)
  * [ ] clusterctl upgrade (clusterctl_upgrade.go, workload flavor: "topology")
  * [ ] Runtime SDK (cluster_upgrade_runtimesdk.go, flavor: "upgrades-runtimesdk")
  * [ ] Self-hosted (self_hosted.go, flavor: "topology")
  * [ ] Autoscaler (autoscaler.go, flavor: "topology-autoscaler")
  * [ ] Dualstack e2e tests
* [ ] Audit if there are any Cluster YAMLs left in the test folder which don't have a machinePool in Cluster.spec.topology

### Misc

* [ ] TBD: internal/controllers/topology/machinepool might be not needed as MachinePool.reconcileDelete deletes templates

### Documentation

* [ ] Update proposal
* [ ] Update docs (probably mostly book)
  * lifecycle hooks
  * builtin variables

### Later

* TBD: consider de-duplicating MD/MP code
  * [Stefan]: Did some deduplication. I personally would keep everything else separate, but open to discussions
* TBD: MachinePool upgrading use Node list instead of Node get
* MHC for MachinePools: depends on MachinePool Machines

/kind feature

**Comment by sbueringer** [comment]
Thx for filing this issue. 
As mentioned in the linked thread, +1 if we don't have any concerns about embedding support for an experimental feature in another experimental feature (which leaks into the non-experimental Cluster resource :)).

Nit regarding the YAML. I would imagine it to look ~ like this:
```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterClass
metadata:
  name: example
spec:
  controlPlane:
...
  workers:
    machinePools:
    - class: example
       template:
          bootstrap:
            ref:
              apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
              kind: KubeadmConfigTemplate
              name: example-pool-0
          infrastructure:
            ref:
              apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
              kind: AzureManagedMachinePool
              name: example-pool-0
---
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: example
  topology:
    class: example
    version: v1.22.4
    ...
    workers:
      machinePools:
      - class: "example"
        name: "mp-0"
        replicas: 10
```

**Comment by fabriziopandini** [comment]
/lifecycle frozen
/help

this is something to complete before moving CC or Machine pools to next phases in graduation

**Comment by killianmuldoon** [comment]
>  Do you have any concerns on this experimental feature relying on another? 

This is already done with ClusterClass and RuntimeSDK so it shouldn't be an issue here IMO.

**Comment by sbueringer** [comment]
I'm fine with it

**Comment by willie-yao** [comment]
@richardcase Thank you so much for working on this! I'm looking at implementing ClusterClass for CAPZ AKS and I'm excited to see this being worked on. If you need any help or just an extra pair of eyes, please let me know and I'll be happy to help out!

**Comment by dtzar** [comment]
@richardcase - really looking forward to this!  Please let us know if we can help.  We'd really like to utilize this in CAPZ.

**Comment by richardcase** [comment]
@dtzar - will do. I had to put it down for a few weeks due to work stuff but picked it up again last week. Hopefully not too much longer until it's complete. We'd also love this in CAPA & CAPG :smile:

**Comment by willie-yao** [comment]
@richardcase Hello! I'm just checking in on how this task is going since you moved the task to the backlog. Please let me know if there's any way I can help!

**Comment by richardcase** [comment]
@willie-yao - I'm struggling to find enough time in a continuous block to make decent progress. Therefore, I'm going to unassign so as not to continue to be a blocker on this.

/remove-lifecycle active
/unassign

**Comment by willie-yao** [comment]
No problem @richardcase! I'll be working on this issue going forward.

/assign

**Comment by sbueringer** [comment]
@willie-yao Let me know if you need some help. I'm pretty familiar with all things ClusterClass.

**Comment by willie-yao** [comment]
Thank you so much @sbueringer!! I will be working off of what Richard has started, but I may need some pointers on where to start as this is my first feature in CAPI. I have already implemented a draft of this feature in CAPZ, but will need to review it since it has been a while: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3197

**Comment by willie-yao** [comment]
@richardcase Can you let me know when you've pushed your existing work to a branch? Thanks so much!

**Comment by sbueringer** [comment]
I talked to Fabrizio and Willie and we would break down this work into the following sub-tasks.

### Base functionality

* [ ] Extend APIs (api/v1beta1)
  * Cluster
    * WorkersTopology: Add MachinePoolTopology slice
  * ClusterClass
    * WorkersClass: Add MachinePoolClass slice
    * PatchSelectorMatch: Add MachinePool
  * Note: excluding MachineHealthCheck fields

* Reconcile MachinePools in Cluster topology (internal/controllers/topology/cluster)
  * [ ] Basic reconcile
    * blueprint.go: also get MachinePool blueprint
    * current_state.go: also get MachinePool current state
    * desired_state.go: calculate desired MachinePool like MachineDeployment
      * This includes version calculation: exact logic is TBD: upgrade sequence, avoid double rollouts in case of concurrent changes, potentially something equivalent to MS preflight checks
    * reconcile_state.go: reconcile MachinePool like MachineDeployment
    * cluster_controller.go: also watch MachinePools
    * Surface MachinePool state like MachineDeployment state in conditions (conditions.go)
  * [ ] Support patching MachinePools:
    * Extend patch engine to also handle MachinePools
    * Extend json patch generator to handle MachinePools
    * Add MachinePool builtin variables
* [ ] TBD: internal/controllers/topology/machinepool might be not needed as MachinePool.reconcileDelete deletes templates

### Validation & Misc

* [ ] Add corresponding validation in internal/webhooks/
  * Includes additions in internal/topology/check/ & internal/topology/variables/{cluster_variable_defaulting.go, cluster_variable_validation.go}

* [ ] Extend internal/controllers/ClusterClass
  * Add the same as we already do for MachineDeployments

### Testing

* [ ] Extend CAPD
  * Add DockerMachinePoolTemplate, maybe more 
* [ ] Extend test/extension
  * Also patch DockerMachinePoolTemplate, maybe more
* [ ] Extend test/e2e
  * Go through current tests and see which ones we should extend for MachinePool (at least the Runtime SDK test) 

### Documentation

* [ ] Update proposal
* [ ] Update docs (probably mostly book)
  * Let's not forget: lifecycle hooks

### Later

* MHC for MachinePools: depends on MachinePool Machines


@dlipovetsky If it's fine for you I would add this to the issue description and use it to track the ongoing work.

**Comment by sbueringer** [comment]
@willie-yao @CecileRobertMichon @fabriziopandini I updated the issue description above to track the follow-up tasks from the current PR. It's mostly about unit test coverage. I went over the PR and identified all relevant files where we have to add unit test coverage.

I think we can basically open one PR per test file. If we have others who want to contribute we can also parallelize this effort once #9016 is merged.

**Comment by fabriziopandini** [comment]
Are we ok with metadata propagation and ownRef management/upgrade (two of the biggest topics in recent releases on top of my mind)?

**Comment by sbueringer** [comment]
We do exactly the same as for MachineDeployments. Except the handling of InfraMachinePool and BootstrapConfig is different, but that is because reconcileDelete in the MP controller takes care of it (i.e. deletes them).

We don't implement any propagation downwards from MachinePool as part of this effort.

**Comment by willie-yao** [comment]
I will be working on the following files for unit tests and create a PR for it shortly:
- internal/controllers/topology/cluster/blueprint_test.go
- internal/controllers/topology/cluster/current_state_test.go
- internal/controllers/topology/cluster/desired_state_test.go

**Comment by johannesfrey** [comment]
If that's ok, I will take a stab at the inline patching tests:

- internal/controllers/topology/cluster/patches/inline/json_patch_generator_test.go
- internal/controllers/topology/cluster/patches/variables/variables_test.go
- internal/controllers/topology/cluster/patches/engine_test.go

**Comment by sbueringer** [comment]
Thx folks, updated the issue description

**Comment by willie-yao** [comment]
I will also be working on tests for these three files:
- internal/controllers/topology/cluster/reconcile_state_test.go
- internal/controllers/topology/cluster/cluster_controller_test.go
- internal/controllers/topology/cluster/conditions_test.go

**Comment by razashahid107** [comment]
Hey everyone
I'd love to work on the following files for unit tests:
- internal/controllers/topology/cluster/scope_test.go
- internal/controllers/topology/cluster/state_test.go
I'll get started working on this and I'll update if I get any questions.

**Comment by sbueringer** [comment]
I assigned you above. Thank you!

**Comment by razashahid107** [comment]
Thank you @sbueringer

**Comment by sbueringer** [comment]
@willie-yao Finally found the time to take a look at our e2e tests. I identified all e2e tests which use ClusterClass and then should eventually also cover MachinePool. I suggest we first start with a PR to cover quickstart. I'm not sure if that will already make it necessary to modify other tests. For now I would try to not parallelize and see how the first few PRs go. WDYT?

**Comment by willie-yao** [comment]
@sbueringer That sounds perfect! Do you mean we shouldn't parallelize the E2E tests with unit tests, or are we good to start working on E2E now?

**Comment by sbueringer** [comment]
@willie-yao We're good to work on it now. Just wouldn't parallelize e2e test PRs. I suspect that once we touch one of them we might have to fix a bunch more (not sure but could happen because we have to change the "topology" flavor which is used in a lot of e2e tests). So parallelizing could lead to duplicate work

**Comment by willie-yao** [comment]
@sbueringer Will be working on **Extend test/e2e**

**Comment by razashahid107** [comment]
Hey @sbueringer. So I was going through scope_test.go to add unit tests and I do not think there is anything to be added in it. I might be mistaken. Can you please verify?

**Comment by chrischdi** [comment]
> Hey @sbueringer. So I was going through scope_test.go to add unit tests and I do not think there is anything to be added in it. I might be mistaken. Can you please verify?

Agree, there's nothing left for `scope_test.go` regards to adding test coverage 👍  we can mark that as resolved / skip that one.

**Comment by sbueringer** [comment]
@willie-yao Updated the e2e test list. #9393 also already fixed the ownerref part, which is great 🎉

**Comment by killianmuldoon** [comment]
Note: It looks like there's a lot of flakiness that has been introduced since adding MachinePools to the ClusterClass self-hosted tests. Those flakes look to be blocking the release currently.

I'm wondering if someone with more expertise in MachinePools is able to take a look: https://github.com/kubernetes-sigs/cluster-api/issues/9522.

Otherwise we might want to remove MPs from the impacted tests ahead of the 1.6 release.

**Comment by sbueringer** [comment]
We could consider disabling them on the release branch to have a clear signal for the release and continue debugging on main?

**Comment by fabriziopandini** [comment]
@sbueringer @willie-yao can we close this issue (eventually moving pending tasks to separate issues if they are a small number, or to a new issue where we can track the Part 2 of this work starting from a clean state)

**Comment by willie-yao** [comment]
> can we close this issue (eventually moving pending tasks to separate issues if they are a small number, or to a new issue where we can track the Part 2 of this work starting from a clean state)

I think that's a good idea! It's a bit harder to track with everything on one issue and all that's left are some unit tests.

**Comment by fabriziopandini** [comment]
/close

@willie-yao will follow up with new issues for what is left

**Comment by fabriziopandini** [comment]
> Not entirely sure what the upside is to move half of an umbrella issue to a new issue

it is just to have a cleaner scope for 1.7 (and possibly an owner of it)
cc @jackfrancis as well
